[DCP] Add `finish_load` hook to `LoadPlanner` for HSDP broadcast replication

niyunsheng opened the pull request:
### Summary

Fixes #171412

This PR introduces a **post-load synchronization hook** to the `LoadPlanner` interface and implements HSDP broadcast replication within `DefaultLoadPlanner`.

Previously, HSDP replicated loading required valid storage reads from all ranks. This PR optimizes this by allowing `DefaultLoadPlanner` to configure a "Leader-Read-Broadcast" pattern via a new `replicate_group` argument.

### Motivation

1. **Mitigate Storage Bottlenecks & Accelerate Loading**
In HSDP setups, all replicas in a replication group need the same data. Standard loading causes every rank to read from storage, leading to linear bandwidth scaling ($N \times \text{Model Size}$).
By shifting the load from the slow file system to the fast internal network (NVLink/IB) via broadcast, we can significantly reduce the I/O bottleneck.

2. **Enforce Atomicity & Clean Abstraction**
**Currently**, achieving this optimization requires a painful workaround in user code: subclassing `LoadPlanner` to filter reads and manually broadcasting tensors *after* `dcp.load()` returns. This approach is dangerous because:
    * **Broken Atomicity:** `dcp.load()` returns with the model in an inconsistent state (only rank 0 has data), risking silent failures if the user forgets to broadcast immediately.
    * **Leaky Abstraction:** Implementation details (NCCL broadcasts) leak into the training loop.

  This PR solves this by internalizing the broadcast logic into `finish_load()`, ensuring `dcp.load()` remains atomic and user code stays clean.

### Implementation Details

1. **`LoadPlanner` Interface Update**:
   - Added a `finish_load()` method to the `LoadPlanner` protocol.
   - This acts as a hook for post-load collective operations. It is guaranteed to be called after storage I/O is complete and all ranks have synchronized at a barrier.

2. **`DefaultLoadPlanner` Enhancements**:
   - Added `replicate_group` argument to `__init__`.
   - **Flexible Broadcasting**: The `replicate_group` can be a **subset** of the full DTensor replication group. This allows fine-grained control over the "Reader-to-Receiver" ratio (e.g., configuring multiple readers per host to balance storage bandwidth vs. network bandwidth).
   - **Validation**: Introduced `_validate_hsdp_replicate_group()` to ensure the provided group is a valid subset of the tensor's physical replication layout.
   - **Plan Generation (`create_local_plan`)**: Filters read requests so that only the **Leader Rank** (Rank 0 of the `replicate_group`) issues storage reads for GPU tensors.
   - **Post-Load Hook (`finish_load`)**: Implements `dist.broadcast` to synchronize loaded data from the leader to other ranks in the input `replicate_group`.

4. **API Cleanup**:
   - Removed the initially proposed `experimental_broadcast_replication` flag from `dcp.load`. This keeps the top-level API surface clean. Users now opt-in via the planner configuration.

5. **CPU Offloading Strategy**:
   - The broadcast optimization is **strictly applied to GPU tensors**.
   - **Reason**: CPU tensors (e.g., optimizer `step`, scalars) are typically negligible in size. Broadcasting them via NCCL usually requires unnecessary host-device synchronization or backend switching, which outweighs the I/O savings. Therefore, small CPU tensors fall back to the standard "all-read" behavior.

### Usage Example

**Basic Usage (Standard HSDP Optimization)**
Configures a single leader to read and broadcast to the entire replication group.
```python
# Old way (Implicit / All-read behavior)
dcp.load(state_dict, checkpoint_id=path)

# New way (HSDP Optimized)
# Rank 0 of the group reads, others receive via broadcast
# The load is atomic; model is fully ready when this returns.
planner = DefaultLoadPlanner(replicate_group=my_hsdp_replicate_group)
dcp.load(state_dict, checkpoint_id=path, planner=planner)
```

**Advanced Usage (Traffic Shaping / Load Balancing)** 
Configures multiple readers within a large replication group by passing a subset process group.
```python
# Example: For a replication group of 8 ranks, split into 2 sub-groups of 4.
# Ranks 0 & 4 read from storage; others receive from their respective local leaders.
# This balances storage I/O bandwidth with NVLink/Network bandwidth.

# sub_group covers ranks [0, 1, 2, 3] or [4, 5, 6, 7]
planner = DefaultLoadPlanner(replicate_group=my_sub_group) 
dcp.load(state_dict, checkpoint_id=path, planner=planner)
```

### Test Plan
- [x] **Enhanced Unit Test**: Updated `test_hsdp_checkpoint.py`.
    - Refactored tests to use `DefaultLoadPlanner(replicate_group=...)` instead of the experimental flag.
    - **Improvement**: Added verification for **Optimizer State** loading (previously, the test only verified model parameters). This ensures the broadcast logic correctly handles replicated optimizer states.
    - Verified that both model weights and optimizer states are restored bitwise-identical across all ranks.

- [x] **Manual Verification**: Confirmed via logging that non-leader ranks do NOT issue storage reads for GPU tensors when the flag is enabled.

### Constraints

- The `replicate_group` passed to the planner must be a **valid subset** of the sharding layout's replication group. (e.g., If the mesh replicates across `[0, 1, 2, 3]`, a user can provide a sub-group `[0, 1]` to broadcast only within that pair).

niyunsheng committed:
distributed checkpoint: add HSDP broadcast loading support

niyunsheng committed:
add TODO: Support mixed meshes

LucasLLC reviewed (changes requested):
Hello! Thanks for submitting a this experimental feature, it seems very useful. My concern is I'm not sure the top level API is the right place for this change. Do we have gaps in the planner interface that prevent us from putting this logic in a custom Planner?

niyunsheng committed:
Refactor: add helper methods for state dict comparisons in HSDP tests

Reduce code duplication by extracting repeated assertion logic into
reusable helper methods.

niyunsheng committed:
Merge branch 'pytorch:main' into feature/hsdp_broadcast_replicated_tensors

niyunsheng committed:
Move HSDP broadcast replication logic from load() to DefaultLoadPlanner

- Add `replicate_group` parameter to DefaultLoadPlanner for HSDP optimization
- Add `finish_load()` method to LoadPlanner interface for post-load collective operations
- Filter broadcast tensors in `create_local_plan()` for non-primary ranks
- Remove `experimental_broadcast_replication` parameter from load() and _load_state_dict()
- Update test to use new planner-based API

niyunsheng committed:
Fix lint issues

niyunsheng committed:
rename items to load

niyunsheng committed:
Move HSDP broadcast replication logic from load() to DefaultLoadPlanner

- Add `replicate_group` parameter to DefaultLoadPlanner for HSDP optimization
- Add `finish_load()` method to LoadPlanner interface for post-load collective operations
- Filter broadcast tensors in `create_local_plan()` for non-primary ranks
- Remove `experimental_broadcast_replication` parameter from load() and _load_state_dict()
- Update test to use new planner-based API

niyunsheng committed:
feat: Add validation for HSDP replicate_group and extend test coverage

- Add `_validate_hsdp_replicate_group()` to verify that the input
  `replicate_group is a subset` of the DTensor's HSDP replicate group
- Extend `test_hsdp_checkpoint` with `broadcast_count` parameter `[0,1,2,3,4]`
  to test multiple broadcast scenarios including sub replicate groups
- Skip tests when `replicate_dim_size` doesn't meet broadcast_count requirements

niyunsheng committed:
Refactor: extract _filter_plan_for_hsdp_broadcast from create_local_plan in DefaultLoadPlanner

niyunsheng committed:
Use coalesced broadcast for HSDP replicated tensors
